ldap/src/main/java/org/springframework/security/ldap/authentication/ad/ActiveDirectoryLdapAuthenticationProvider.java

```diff
@@ -398,8 +408,16 @@ public void setSearchFilter(String searchFilter) {
 		this.searchFilter = searchFilter;
 	}
 
-	static class ContextFactory {
-		DirContext createContext(Hashtable<?, ?> env) throws NamingException {
+	/**
+	 * Factory for directory context
+	 */
+	public static class ContextFactory {
```

Comment:
Rather than exposing a custom type, I think we should change this type to `ContextSource`.

Reply:
What do you think about using (and exposing via ```setContextFactory```) less generic ```LdapContextSource```?
It is perfectly suite AD provider needs as I think.

Reply:
We should not create a new public API unless we really need it.